projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php

```diff
@@ -10,24 +10,95 @@ class Jetpack_Widget_Conditions {
 	static $passed_template_redirect = false;
 
 	public static function init() {
-		if ( is_admin() ) {
-			add_action( 'sidebar_admin_setup', array( __CLASS__, 'widget_admin_setup' ) );
-			add_filter( 'widget_update_callback', array( __CLASS__, 'widget_update' ), 10, 3 );
+		global $pagenow;
+
+		// Previews don't need widget editing loaded and also don't want to run the filter - it might mean the preview
+		// is empty which would be confusing.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( isset( $_GET['legacy-widget-preview'] ) ) {
+			return;
+		}
+
+		// Not making use of the heartbeat functionality at all.
+		// phpcs:ignore WordPress.Security.NonceVerification.Missing
+		if ( isset( $_POST['action'] ) && 'heartbeat' === $_POST['action'] ) {
+			return;
+		}
```

Comment:
This confused me a bit. Would it be possible to target a specific AJAX request (`save-widget`?) and bail if it's anything but that?

Reply:
Do you mean if action is set and it's not save-widget then bail? That would probably be fine.

I added this guard clause as this functionality isn't used with the [heartbeat API](https://developer.wordpress.org/plugins/javascript/heartbeat-api/), so this will prevent some hooks from being registered. As the hooks will not get ran anyway, it's probably a bit of a premature optimisation and can be dropped to avoid confusion.